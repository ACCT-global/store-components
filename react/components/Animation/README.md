# Animation

## Description

`Animation` is a VTEX Component that display some animations in their children props. 
This Component can be imported and used by any VTEX App.

:loudspeaker: **Disclaimer:** Don't fork this project, use, contribute, or open issue with your feature request.

## Table of Contents
- [Usage](#usage)

## Usage

To import it into your code: 
```js
import Animation from 'vtex.store-components/Animation'
```

You can use it in your code like a React component with the jsx tag: `<Animation />`. 
```jsx
<Animation> 
  This is an animated text! \o/
</Animation>
```

| Prop name          | Type       | Description                                                                                 |
| ------------------ | ---------- | ------------------------------------------------------------------------------------------- |
| `className`        | `String`   | The className to set the animation                                                          |
| `children`         | `Node!`    | Component children that will be displayed inside of the animation                           |
| `type`             | `String`   | Type of the animation, it can be 'drawerLeft', 'drawerRight', 'drawerTop' or 'drawerBottom' |
| `isActive`         | `Boolean`  | Active the animation                                                                        |
| `duration`         | `number`   | The animation duration in seconds, the default value is 0.4 second.                         |
| `transfer`         | `number`   | The active animation deslocation in percentage, the default value is 110%.                  |
| `transferEnter`    | `number`   | The not active animation deslocation in percentage, the default value is 0%.                |