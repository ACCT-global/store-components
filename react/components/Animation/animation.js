export const DRAWER_ANIMATION = {
  drawerHorizontal: (duration, transfer) => ({
    transform: `translateX(${transfer}%)`,
    transition: `transform ${duration}s ease-in-out`,
  }),
  drawerVertical: (duration, transfer) => ({
    transform: `translateY(${transfer}%)`,
    transition: `transform ${duration}s ease-in-out`,
  }),
}

export const ANIMATIONS = {
  drawerLeft: {
    from: (duration, transfer) => DRAWER_ANIMATION['drawerHorizontal'](duration, transfer - transfer),
    leave: (duration, transfer) => DRAWER_ANIMATION['drawerHorizontal'](duration, transfer),
  },
  drawerRight: {
    from: (duration, transfer) => DRAWER_ANIMATION['drawerHorizontal'](duration, transfer - transfer),
    leave: (duration, transfer) => DRAWER_ANIMATION['drawerHorizontal'](duration, -transfer),
  },
  drawerTop: {
    from: (duration, transfer) => DRAWER_ANIMATION['drawerVertical'](duration, transfer - transfer),
    leave: (duration, transfer) => DRAWER_ANIMATION['drawerVertical'](duration, transfer),
  },
  drawerBottom: {
    from: (duration, transfer) => DRAWER_ANIMATION['drawerVertical'](duration, transfer - transfer),
    leave: (duration, transfer) => DRAWER_ANIMATION['drawerVertical'](duration, -transfer),
  },
}
