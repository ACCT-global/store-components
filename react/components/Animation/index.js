import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { Transition } from 'react-spring'

const ANIMATIONS = {
  drawerLeft: {
    from: { transform: 'translateX(100%)' },
    enter: { transform: 'translateX(0%)' },
    leave: { transform: 'translateX(100%)' },
  },
  drawerRight: {
    from: { transform: 'translateX(-100%)' },
    enter: { transform: 'translateX(0%)' },
    leave: { transform: 'translateX(-100%)' },
  },
  drawerTop: {
    from: { transform: 'translateY(-100%)' },
    enter: { transform: 'translateY(0%)' },
    leave: { transform: 'translateY(-100%)' },
  },
  drawerBottom: {
    from: { transform: 'translateY(100%)' },
    enter: { transform: 'translateY(0%)' },
    leave: { transform: 'translateY(100%)' },
  },
}

/**
 * Animation component
 */
export default class Animation extends Component {
  static propTypes = {
    /* Object to be animated */
    children: PropTypes.object.isRequired,
    /* Active the animation */
    isActive: PropTypes.bool,
    /* Classname to the animation */
    className: PropTypes.string,
    /* Type of animation */
    type: PropTypes.oneOf(['drawerLeft', 'drawerRight', 'drawerTop', 'drawerBottom']),
  }

  static defaultProps = {
    className: '',
    type: 'drawerLeft',
  }

  renderChildren = style => (
    <div className={this.props.className}
      style={style}
    >
      {this.props.children}
    </div>
  )

  render() {
    const { isActive, type } = this.props
    const style = ANIMATIONS[type]

    return (
      <Transition
        keys={isActive ? ['children'] : []}
        {...style}
      >
        {isActive ? [this.renderChildren] : []}
      </Transition>
    )
  }
}
