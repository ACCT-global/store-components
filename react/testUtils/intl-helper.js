import React from 'react'
import * as reactTestingLibrary from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import defaultStrings from '../../messages/en-US.json'
const Context = React.createContext({})

export const render = (node, options) => {
  const rendered = reactTestingLibrary.render(
    <IntlProvider messages={defaultStrings} locale="en-US">
      {node}
    </IntlProvider>,
    options
  )

  return {
    ...rendered,
    rerender: newUi =>
      render(newUi, {
        container: rendered.container,
        baseElement: rendered.baseElement,
      }),
  }
}

// re-export everything
export default reactTestingLibrary
