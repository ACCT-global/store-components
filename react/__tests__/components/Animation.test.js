import React from 'react'

import Animation from '../../Animation'
import { renderWithIntl } from '../../__helpers__/intl-helper'

describe('<Animation /> component', () => {
  let wrapper

  const renderComponent = customProps => {
    const comp = <Animation {...customProps}> Test </Animation>

    return renderWithIntl(comp)
  }

  beforeEach(() => {
    wrapper = renderComponent()
  })

  it('should be rendered', () => {
    expect(wrapper).toBeDefined()
    expect(wrapper).toBeTruthy()
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
