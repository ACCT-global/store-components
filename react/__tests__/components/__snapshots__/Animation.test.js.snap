// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Animation /> component should match snapshot 1`] = `
<div
  class=""
  style="transform:translateX(110%);transition:transform 0.4s ease-in-out"
>
   Test 
</div>
`;
